### PR TITLE
Year of week-of-year support

### PR DESCRIPTION
**Pull request by mildgravitas**
Implements year of year-of-week support for #488 and:
-  Adds a dedicated Year enum to datetime::options::components::Bag.
- Loosens get_best_available_format_pattern() & adjust_pattern_field_lengths() to match FieldSymbol enums but not their data. This is necessary for components::Year matching to work & also improves h12<->h23 coerced time patterns which previously dropped the timezone.

### CODE REVIEW

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs:348]: For compatibility with existing usage in test fixtures. I can remove the renames & update the latter if preferable.

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs:358]: They're not equivalent but the difference is subtle. I've added examples to make it more salient.

File: provider/testdata/data/json/datetime/gregory_lengths@1/ar-EG.json
Comment (mildgravitas) [reply provider/testdata/data/json/datetime/gregory_lengths@1/ar-EG.json:16]: Ah, that's a consequence of the `discriminant_cmp()` changes to skeleton matching: the changed patterns are 12<->24 hour translations of CLDR's full or long time patterns. The algorithm (from components/datetime/src/pattern/hour_cycle.rs) here is:

1. Take the original pattern & swap 12<->24 hour symbols. e.g. 'h:mm:ss a zzzz' -> 'H:mm:ss a zzzz'.
2. Turn the pattern into a skeleton. e.g. 'H:mm:ss a zzzz' -> 'Hmsv'.
3. Call `skeleton::create_best_pattern_for_fields()` to find a matching pattern.

Previously 'v' & 'z' were considered distinct so `create_best_pattern_for_fields()` would settle for a partial match of 'Hmsv' against 'Hms'. Now it finds an exact match.

It's possible to cancel this change by altering `discriminant_cmp()` to compare the inner data for `TimeZone`, but given that this is arguably more correct I left it in.

**Comment by mildgravitas** [comment]
Rebased onto dc5ff31e24e1203d719701cb179b3a6f3df3eac0

File: components/datetime/src/options/components.rs
Comment (zbraniecki) [reply components/datetime/src/options/components.rs:348]: please do. ICU4X data shouldn't use renames to conform to anything external. And test fixtures should be updated when the values change.

File: components/datetime/src/options/components.rs
Comment (zbraniecki) [reply components/datetime/src/options/components.rs:358]: `nit`: Thanks, that's helpful, but the reader still has to derive an implicit information about the actual value. Could you extend the comments for `NumericWeekOf` and `TwoDigitWeekOf` to:

```
    /// The numeric value of the year in "week-of-year", such as "2019" of the "2019-01" for the
    /// week of 2018-12-31 according to the ISO calendar.
```

**Review by zbraniecki — APPROVED** [review]

**Comment by zbraniecki** [comment]
This looks good to me - with the changes I requested.

I'd like @nordzilla or @gregtatum to review the skeleton's changes in particular before we merge this.

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs:356]: Done

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs]: Done

**Comment by mildgravitas** [comment]
Rebased onto 0c855b106c6e8b8a66142defc6102885ccc8c7c5

File: components/datetime/src/skeleton/helpers.rs
Comment (nordzilla) [inline components/datetime/src/skeleton/helpers.rs:302]: > nipick (style | optional)

Rust 1.53 introduced the [`is_eq()`](https://doc.rust-lang.org/stable/std/cmp/enum.Ordering.html#method.is_eq) function for `Ordering`. 

I'd like to see this be 

```rust
field.symbol.discriminant_cmp(&pattern_field.symbol).is_eq()
```

File: components/datetime/benches/fixtures/tests/components.json
Comment (nordzilla) [inline components/datetime/benches/fixtures/tests/components.json:27]: > thought

I would prefer to see uniform casing in the JSON data, rather than a mixture of`"Numeric"` and `"numeric"`, but we can fix this. See the suggestion comment where the `Year` enum is defined.

File: components/datetime/src/options/components.rs
Comment (nordzilla) [inline components/datetime/src/options/components.rs:345]: > suggestion

We should add

```rust
#[serde(rename_all = "camelCase")]
```

This should fix the inconsistent JSON data that contains both `"Numeric"` and `"numeric"`.

**Review by nordzilla — CHANGES_REQUESTED** [review]
Looks good to me overall, though @gregtatum  is certainly more of an authority on the skeletons and patterns here, so I'd still like to wait for his review. 

I had a couple things I'd like to see changed with regard to style.

File: components/datetime/src/options/components.rs
Comment (zbraniecki) [reply components/datetime/src/options/components.rs:345]: Ah, I see my mistake. @gregtatum did already add renaming for option components and @mildgravitas was just following that, and I was unaware because the diff only showed the additions of renames.

I think I'd generally prefer not to rename anything unnecessairly, but in this case, I can imagine users wanting to react options bags from JSON, so I'd be okay to use renames.

My only recommendation would be to use `kebab-case` for all of them - @mildgravitas can you update all of component enums (including the one you're adding) to use `kebab-case`?

**Comment by zbraniecki** [comment]
@mildgravitas - that looks good to me! Can you please move all components to use `kebab-case`? including the ones that are already there now with camel case (like Numeric, Text, Month, Week, TimeZoneName)?

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs]: Done: I've added rename_all = kebab-case to the Year & Week enums.

File: components/datetime/benches/fixtures/tests/components.json
Comment (mildgravitas) [reply components/datetime/benches/fixtures/tests/components.json]: Done

File: components/datetime/src/skeleton/helpers.rs
Comment (mildgravitas) [reply components/datetime/src/skeleton/helpers.rs]: Done

File: components/datetime/src/options/components.rs
Comment (mildgravitas) [reply components/datetime/src/options/components.rs]: And the same for all other component enums.

**Review by zbraniecki — DISMISSED** [review]
lgtm! thank you!

**Review by nordzilla — APPROVED** [review]
LGTM, thank you!

File: components/datetime/tests/fixtures/tests/components-width-differences.json
Comment (gregtatum) [inline components/datetime/tests/fixtures/tests/components-width-differences.json:51]: Suggestion (optional): It seems possible to create strange combinations of week of year and other options in the components bag. It would be good to exercise these code paths with more robust tests. If we're not getting good results it might be worth filing bugs for them, or having `should_panic` tests.

File: provider/testdata/data/json/datetime/gregory_lengths@1/ar-EG.json
Comment (gregtatum) [reply provider/testdata/data/json/datetime/gregory_lengths@1/ar-EG.json:16]: Thanks for the fix!

**Review by gregtatum — DISMISSED** [review]
The algorithm changes look good to me, and fix some existing issues. I left a few suggestions on more testing configurations for locales and combinations, but made them non-blocking. Sorry for the delay on this review, I've been out of office for a week.

**Comment by mildgravitas** [comment]
Rebased onto e12752342b684bf6e19ce1b7157b5bb24090422f

**Review by zbraniecki — APPROVED** [review]

**Comment by zbraniecki** [comment]
@mildgravitas there are two optional suggestions from @gregtatum left. Do you want to apply them or land-as-is?

**Comment by mildgravitas** [comment]
Apologies for the delay.

Rebased onto 0845aa18ff40b5f7a0dfad8b7d63ec86b55d3d59.

File: components/datetime/tests/fixtures/tests/components-width-differences.json
Comment (mildgravitas) [reply components/datetime/tests/fixtures/tests/components-width-differences.json]: Done

File: components/datetime/tests/fixtures/tests/components-width-differences.json
Comment (mildgravitas) [reply components/datetime/tests/fixtures/tests/components-width-differences.json:55]: I've added a components-partial-matches fixture with a few combinations. This revealed that empty matches (e.g. by only asking for the week field) result in an empty string being rendered so I changed that case to return an error instead.

I also noticed that 'm' becomes 'm:ss' rather than failing to match. This is also the case upstream & the result is functional so I've just documented it with a test.

**Comment by Manishearth** [comment]
@zbraniecki @gregtatum are there any leftover things to do to land this PR? I think we're just missing a final review.

**Comment by Manishearth** [comment]
@mildgravitas btw, we changed how the data is stored, so if you rebase again you probably will want to `cargo make download-testdata` and then `cargo make testdata` to generate the new data

**Comment by mildgravitas** [comment]
@Manishearth: Thanks for the heads up. Seems like I'm already top-of-tree from yesterday's rebase onto dc414a8fcaeafe0625c576d90adb8053cdfca7fa. Running testdata-download && testdata produces no diffs so all good for the time being it seems.

**Comment by Manishearth** [comment]
Ah! I was looking at an older force-push-webhook comment which listed `gregory_lengths` etc. Yep, your latest push should be fine!

**Review by gregtatum — APPROVED** [review]
It's looking good to me! Thanks for addressing my comments.

**Comment by Manishearth** [comment]
Going to hit merge since Zibi already gave an "r+ except for minor issues" review above and they seem to have been addressed; I'd rather not let this bitrot again :smile:

**Pull request merged**
